### PR TITLE
os/fs/smartfs : Apply Basic Step of Journal

### PR DESCRIPTION
**Pull request by Taejun-Kwon**
1. Apply Logic regarding journal block
2. Apply checkout / checkin logic
3. Optimize Bit-operation of journal type and state
4. Apply basic logic of init and recovery

### CODE REVIEW

File: os/fs/driver/mtd/smart.c [hunk 1]
```diff
@@ -4908,262 +4933,495 @@ static int smart_ioctl(FAR struct inode *inode, int cmd, unsigned long arg)
...
+	case SMART_JOURNAL_TYPE_COMMIT: 
+		break;
+	case SMART_JOURNAL_TYPE_RELEASE:
+		break;
+	case SMART_JOURNAL_TYPE_ERASE:
+		break;
+	default:
+		break;
+	}
+
```
Comment (jsdosa) [inline os/fs/driver/mtd/smart.c:5422]: could you remove blank line?

Comment (Taejun-Kwon) [reply os/fs/driver/mtd/smart.c:5422]: okay.

File: os/fs/driver/mtd/smart.c [hunk 2]
```diff
@@ -4908,262 +4933,495 @@ static int smart_ioctl(FAR struct inode *inode, int cmd, unsigned long arg)
...
- * Name: smart_journal_bytewrite
+ * Name: smart_journal_scan
  *
  * Description:
- *   Byte Write function instead of MTD_WRITE
- *   It is called when smartfs release or alloc target sector, it means no data.
- *   This will not write data on physical sector.
- *   Sequence is journal write -> write header
+ *   Scan Entire of journal blocks and set dev->journal_seq.
+ *   Recovery which hass checkin state, but not checkout state.
```
Comment (amandeepc) [inline os/fs/driver/mtd/smart.c:5324]: *has

File: os/fs/driver/mtd/smart.c [hunk 3]
```diff
@@ -4908,262 +4933,508 @@ static int smart_ioctl(FAR struct inode *inode, int cmd, unsigned long arg)
...
+		break;
+
+	/* TODO There's chance that add verify code for below cases. If it is not necessary later, Please Remove them! */			
+	case SMART_JOURNAL_TYPE_ERASE:
+		/** If erase need to be check in future, verify logic should added.
+		  * But I think It had better done by MTD Driver itself to reduce delay..
+		  */
+		break;
+
+	/* Release type, we don't have to nothing for now, but if it needed, we can check written data of mtd header */
```
Comment (sunghan-chang) [inline os/fs/driver/mtd/smart.c:5106]: Is it `we have to nothing`?

**Pull request merged**